isaaceindhoven/testcafe-reporter-allure
# feat(reporter): Added support for "Tags" labels

Hi,
In this change I would like to submit implementation for tags meta support. 

**Description of the Change**

allure-js-common supports adding multiple labels with value `tag`:

```
export declare enum LabelName {
    AS_ID = "AS_ID",
    SUITE = "suite",
    PARENT_SUITE = "parentSuite",
    SUB_SUITE = "subSuite",
    EPIC = "epic",
    FEATURE = "feature",
    STORY = "story",
    SEVERITY = "severity",
    TAG = "tag",
    OWNER = "owner",
    LEAD = "lead",
    HOST = "host",
    THREAD = "thread",
    TEST_METHOD = "testMethod",
    TEST_CLASS = "testClass",
    PACKAGE = "package",
    FRAMEWORK = "framework",
    LANGUAGE = "language"
}
```
Multiple tags values could be shown in allure report on testcase view: 

<img width="529" alt="Screen Shot 2021-11-25 at 14 52 16" src="https://user-images.githubusercontent.com/2556875/143410920-cae98764-c19a-48ef-83a1-07acf5223779.png">

In my solution I implementing accepting custom metatag `tags` that accept **array of strings** or **string** value.

**First example with `tags` as and array**
```
test
  .meta({
    severity: 'trivial',
    issue: 'TEST-ISSUE',
    description: 'An example description',
    epic: 'Example Epic Ticket',
    feature: 'Example Feature Ticket',
    story: 'Example Story Ticket',
    suite: 'Main Example Group',
    tags: ['e2e', 'not-automated'],
    skipReason: 'Skipped: Not automated',
    // ... any other key: value property as custom metadata
  })
  .skip('Example planned test', async () => {
    // Test Code
  });

```

First example would produce following allure raw result: 
```
{
  "uuid": "8cef46ab-3a89-49ab-ac31-beb5d2aa7249",
  "historyId": "0d9eda76-ff1f-4892-86b7-7d8f0581290e",
  "status": "skipped",
  "statusDetails": { "message": "Skipped: Not automated", "trace": "" },
  "stage": "finished",
  "steps": [],
  "attachments": [],
  "parameters": [{ "name": "skipReason", "value": "Skipped: Not automated" }],
  "labels": [
    { "name": "framework", "value": "TestCafe" },
    { "name": "language", "value": "typescript/javascript" },
    { "name": "severity", "value": "trivial" },
    { "name": "suite", "value": "TestCafé Example Fixture - Documentation" },
    { "name": "subSuite", "value": "Main Example Group" },
    { "name": "epic", "value": "Example Epic Ticket" },
    { "name": "feature", "value": "Example Feature Ticket" },
    { "name": "story", "value": "Example Story Ticket" },
    { "name": "tag", "value": "e2e" },
    { "name": "tag", "value": "not-automated" }
  ],
  "links": [{ "name": "TestRail: TEST-ISSUE", "url": "https://jira.example.nl/browse/TEST-ISSUE", "type": "issue" }],
  "start": 1637821538338,
  "name": "Example planned test",
  "fullName": "TestCafé Example Fixture - Documentation : Example planned test",
  "description": "An example description",
  "stop": 1637821538339
}
```
First example will show two tags in allure results:

<img width="1335" alt="firstExample" src="https://user-images.githubusercontent.com/2556875/143394661-4dcfbc17-4efa-42a2-be16-126b0a371288.png">

**Second example with `tags` as a string**
It is possible to pass `tags` as string of one tag that will be converted in array with one element.
Example TestCafe tests with `tags` value passed as string: 
```
test
  .meta({
    suite: 'Another Example Group',
    tags: 'not-automated', // string tags are also supported
    skipReason: 'Not automated',
  })
  .skip('Example not automated e2e test 3', async () => {
    // planned test
  });
```
Second example would produce following allure raw result: 
```
{
  "uuid": "de5c0baf-c5c2-4c14-8d79-8119c0233c10",
  "historyId": "4e7d5756-66aa-4f79-87c6-9c172e2b1850",
  "status": "skipped",
  "statusDetails": { "message": "Not automated", "trace": "" },
  "stage": "finished",
  "steps": [],
  "attachments": [],
  "parameters": [{ "name": "skipReason", "value": "Not automated" }],
  "labels": [
    { "name": "framework", "value": "TestCafe" },
    { "name": "language", "value": "typescript/javascript" },
    { "name": "severity", "value": "Normal" },
    { "name": "suite", "value": "TestCafé Example Fixture - Documentation" },
    { "name": "subSuite", "value": "Another Example Group" },
    { "name": "tag", "value": "not-automated" }
  ],
  "links": [],
  "start": 1637823082517,
  "name": "Example not automated e2e test 3",
  "fullName": "TestCafé Example Fixture - Documentation : Example not automated e2e test 3",
  "stop": 1637823082518
}
```
Second example will show one tag in allure results:

<img width="1343" alt="second example" src="https://user-images.githubusercontent.com/2556875/143394836-e3205205-b31f-46b0-a18c-ed0001b8d958.png">

**Alternate Designs**
I think this solution is the most allure-native.

**Verification Process**

> How did you verify that all new functionality works as expected?
Added two simple tests in `examples/base-implementation-with-custom-config/tests/docMetaExample.spec.ts`

1. Run `cd examples/base-implementation-with-custom-config`
2. Run `npm run test:doc:example` > tests passed and created raw allure results in `examples/base-implementation-with-custom-config/allure-results`
3. Run `npm run allure` > command opened allure report where I verified that both examples produces tags section on test overview. Added snapshot in documentation.

Actual result after steps:

<img width="1335" alt="firstExample" src="https://user-images.githubusercontent.com/2556875/143413203-41fa9815-fba4-4549-b4ae-e974bc325c4f.png">

> How did you verify that the change has not introduced any regressions?

1. Added two simple tests to include two test cases: `tags` with `string` value and `tags` with `string[]`
2. Run `npm run test:e2e`, then `npm run test:integration -- -u` and updated snapshost. 
3. Verified snapshot is updated and not other changes introduced

**Additional notes**
I have also updated documentation with examples on how to use `tags`.
